consensus/src/ledger/merkle.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+
+use crate::Ledger;
+use anyhow::*;
+use indexmap::IndexSet;
+use snarkos_storage::Digest;
+use snarkvm_algorithms::MerkleParameters;
+
+use super::indexed_merkle_tree::IndexedMerkleTree;
+
+#[derive(Clone)]
```

Comment:
Should this be deep-clonable?

Reply:
it is deep clonable? yes. this is somewhat of a performance issue in the future, but nothing observed as a problem so far. we clone for transactionality/revertability, but in the future we should use a better model. this is fast and notable doesn't need any snarkvm changes.

Reply:
I'm ok with this as a temporary measure, but these collections will grow with time, won't they? It would be best to put a `FIXME` or `TODO` here with the rationale behind the deep cloning and possible future solutions.